Relax nanosecond datetime restriction in CF time decoding

kmuehlbauer opened the pull request:
<!-- Feel free to remove check-list items aren't relevant to your change -->

- [x] Closes #7493
- [x] Closes #2587
- [x] Tests added/changed
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] New functions/methods are listed in api.rst

This is another attempt to resolve #7493. This goes a step further than #9580.

The idea of this PR is to automatically infer the needed resolutions for decoding/~encoding~ and only keep the constraints pandas imposes ("s" - lowest resolution, "ns" - highest resolution). There is still the idea of a `default resolution`, but this should only take precedence if it doesn't clash with the automatic inference. This can be discussed, though. Update: I've implemented `time-unit`-kwarg ~a first try to have default resolution~ on decode, which will override the current inferred resolution only to higher resolution (eg. `'s'` -> `'ns'`). To work towards #4490 the time decoding options (`decode_time` and `use_cftime` are bundled within `CFDatetimeCoder` which is distributed via `decode_times` kwarg. `use_cftime`-kwarg is deprecated.

For sanity checking, and also for my own good, I've created [a documentation page on time-coding in the internal dev section](https://xray--9618.org.readthedocs.build/en/9618/internals/time-coding.html). Any suggestions (especially grammar) or ideas for enhancements are much appreciated.

There still might be room for consolidation of functions/methods (mostly in coding/times.py), but I have to leave it alone for some days. I went down that rabbit hole and need to relax, too :grimacing:.

Looking forward to get your insights here, @spencerkclark, @ChrisBarker-NOAA, @pydata/xarray.

Todo:

- [x] floating point handling
- [x] update decoding tests to iterate over `time_units` (where appropriate)
- [x] `CFDatetimeCoder` as input for `decode_times` kwarg
- [ ] ...

kmuehlbauer commented:
Nice, mypy 1.12 is out and breaks our typing, :sob:.

TomNicholas commented:
> Nice, mypy 1.12 is out and breaks our typing, 😭

Can we pin it in the CI temporarily?

kmuehlbauer commented:
> Can we pin it in the CI temporarily?

Yes, 1.11.2 was the last version.

kmuehlbauer commented:
@spencerkclark Thanks for taking the time to review. I've tried to be explicit with comments and maybe also my doc ramblings will help. I've also tried to fixup the current docs on time handling. 

There are still some issues when creating datetime/timedelta from scratch which currently flow finally through pd.Series if not converted by other means beforehand. So, there is still some glitches which have to be ironed out.

Anyway, thanks and good luck!

spencerkclark reviewed (commented):
Thanks @kmuehlbauer—I think this is coming together well.  I'll provide a more detailed review eventually.  In my mind there are still some high-level things to consider:

- To keep breaking changes from happening abruptly, we may want to have a period where the default decoding `time_resolution` remains `"ns"` before changing to something else.  From my initial scan of how you modified the tests—maybe with foresight—I think it would be straightforward to switch back to defaulting to `"ns"` precision, which is nice.  However, we still would want a way to test non-nanosecond precision behavior.  Is this what you have been thinking about or did you have something else in mind?  I'm happy to hear other points of view.  At least in the case of casting in the Variable constructor we have been warning about this for a while.
- How to best handle choosing a precision when decoding times encoded as floats is still an open question.  I left a couple thoughts below, but I think it may need more discussion.
- We should think more about how we handle casting when constructing Variable objects with `datetime.datetime` or `datetime.timedelta` objects—our current approach may not be up to the task for objects that cannot be represented with nanosecond precision, though this may be considered an upstream issue.

Something else to keep on our radar is the discussion in https://github.com/pandas-dev/pandas/issues/58989.  It looks like pandas is reconsidering how to handle precision inference in various contexts—I think what they are leaning towards is pretty similar to what you have currently for decoding, which is good.

spencerkclark commented on xarray/coding/times.py:
```diff
@@ -479,17 +560,17 @@ def cftime_to_nptime(times, raise_on_invalid: bool = True) -> np.ndarray:
-    for i, t in np.ndenumerate(times):
+    for _i, t in np.ndenumerate(times):
         try:
+            # todo: decide how to work with this
+            #  as initialized by string pd.Timestamp is defined only from year -9999-01-01 to 9999-12-31
```
Oh I didn't realize this actually fails too:
```
>>> pd.Timestamp(10000, 1, 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "timestamps.pyx", line 1850, in pandas._libs.tslibs.timestamps.Timestamp.__new__
ValueError: year 10000 is out of range
```

kmuehlbauer replied:
This moves probably through the same code path ("string" -> (years, months, days, etc) -> Timestamp()).

This is a little bit annoying, but given the specialties I'll not blame pandas in any way. The move to non-nanosecond resolution was a major undertaking and issues will be ironed out as they appear, I think.

Yes, we can try to as you suggest.

kmuehlbauer commented on doc/internals/time-coding.rst:
```diff
@@ -0,0 +1,442 @@
+    print("DatetimeIndex as_unit('s'):", time.as_unit("s"))
+    print("DatetimeIndex to_numpy():", time.as_unit("s").to_numpy())
+
+.. warning::
+    Input data with resolution higher than ``'ns'`` (eg. ``'ps'``, ``'fs'``, ``'as'``) is truncated (not rounded) at the ``'ns'``-level. This is currently broken for the ``'ps'`` input, where it is interpreted as ``'ns'``.
```
I'll create an issue on the pandas tracker, when I'm back next week.

kmuehlbauer commented on doc/internals/time-coding.rst:
```diff
@@ -0,0 +1,442 @@
+            np.array([np.datetime64(1901901901901, "s")]),
+            pd.to_datetime(np.array([np.datetime64(1901901901901, "s")])),
+        )
+        try:
+            pd.to_datetime([np.datetime64(1901901901901, "s")])
+        except Exception as err:
+            print("Raises:", err)
+        try:
+            pd.to_datetime(1901901901901, unit="s")
+        except Exception as err:
+            print("Raises:", err)
+        try:
+            pd.to_datetime([1901901901901], unit="s")
+        except Exception as err:
+            print("Raises:", err)
+        try:
+            pd.to_datetime(np.array([1901901901901]), unit="s")
+        except Exception as err:
+            print("Raises:", err)
```
Yes, this was more of documenting this for me, because I've tripped over this quite some time ;-)

kmuehlbauer commented on doc/internals/time-coding.rst:
```diff
@@ -0,0 +1,442 @@
+Since relaxing the resolution this enhances the range to several hundreds of thousands of centuries with microsecond representation. ``NaT`` will be at ``np.iinfo("int64").min`` for all of the different representations.
+
+.. warning::
+    When initialized with a datetime string this is only defined from ``-9999-01-01`` to ``9999-12-31``.
```
Yes, so it seems. I tried, but no neat solution came to mind.

kmuehlbauer replied:
I think we can utilize the enhanced iso8601-parser (#9885) to parse the ref date:

```python
ref_date = "10000-01-01T00:00:00.1"
xr.coding.cftimeindex._parse_iso8601_without_reso(pd.Timestamp, ref_date)
```
```python
Timestamp('10000-01-01 00:00:00.100000')
```

We might even think to enhance this to nanosecond parsing. I'll give this a try.

kmuehlbauer commented on doc/internals/time-coding.rst:
```diff
@@ -0,0 +1,442 @@
+
+2. ``standard``/``gregorian`` calendar and ``proleptic_gregorian`` are equivalent for any dates and reference times >= ``1582-10-15``. First the reference time is checked and any timezone information stripped off and in a second step, the minimum and maximum ``values`` are checked if they can be represented in the current reference time resolution. At the same time integer overflow would be caugh ...
+
+3. As the time unit (here ``seconds``) and the resolution of the reference time ``1992-10-8 15:15:42.5 -6:00`` (here ``milliseconds``) might be different, this has to be aligned to the higher resolution (retrieve new unit). This is done by multiplying the ``values`` by the ratio of nanoseconds per time unit and nanoseconds per reference time unit. To not break consistency for ``NaT`` a mask is ke ...
```
Yes, special casing floats, also another parameter as suggested is a good step forward. I'll think about that more.

kmuehlbauer replied:
With the latest incarnation of floating point handling this works now as follows:

```python
>>> values = np.array([0, 0.25, 0.5, 0.75, 1.0], dtype="float64")
>>> units = "seconds since 2000-01-01"
>>> calendar = "standard"
>>> xr.coding.times.decode_cf_datetime(values, units, calendar)
array(['2000-01-01T00:00:00.000000000', '2000-01-01T00:00:00.250000000',
       '2000-01-01T00:00:00.500000000', '2000-01-01T00:00:00.750000000',
       '2000-01-01T00:00:01.000000000'], dtype='datetime64[ns]')
```

If you need a dedicated time_unit:

```python
>>> values = np.array([0, 0.25, 0.5, 0.75, 1.0], dtype="float64")
>>> units = "seconds since 2000-01-01"
>>> calendar = "standard"
>>> xr.coding.times.decode_cf_datetime(values, units, calendar, time_unit="ms")
array(['2000-01-01T00:00:00.000', '2000-01-01T00:00:00.250',
       '2000-01-01T00:00:00.500', '2000-01-01T00:00:00.750',
       '2000-01-01T00:00:01.000'], dtype='datetime64[ms]')
```

If your floats can't be represented, we try to overcome this (note the "days"):

```python
>>> values = np.array([0, 0.00025, 0.0005, 0.00075, 1.0], dtype="float64")
>>> units = "days since 2000-01-01"
>>> calendar = "standard"
>>> xr.coding.times.decode_cf_datetime(values, units, calendar, time_unit="s")
/tmp/ipykernel_7190/3938849960.py:4: SerializationWarning: Can't decode floating point datetime to 's' without precision loss, decoding to 'ms' instead. To silence this warning use time_unit='ms' in call to decoding function.
  xr.coding.times.decode_cf_datetime(values, units, calendar, time_unit="s")

array(['2000-01-01T00:00:00.000', '2000-01-01T00:00:21.600',
       '2000-01-01T00:00:43.200', '2000-01-01T00:01:04.800',
       '2000-01-02T00:00:00.000'], dtype='datetime64[ms]')
```

kmuehlbauer replied:
Hope this resolves the floating point issues.

kmuehlbauer commented on doc/user-guide/weather-climate.rst:
```diff
@@ -137,7 +132,7 @@ Conversion between non-standard calendar and to/from pandas DatetimeIndexes is
 :py:meth:`xarray.DataArray.convert_calendar`). Here, like elsewhere in xarray, the ``use_cftime``
 argument controls which datetime backend is used in the output. The default (``None``) is to
-use `pandas` when possible, i.e. when the calendar is standard and dates are within 1678 and 2262.
+use `pandas` when possible, i.e. when the calendar is standard and dates starting with 1582-10-15.
```
Yes, this is only a limitation for `standard`/`gregorian`. This needs change.

kmuehlbauer replied:
added a note

kmuehlbauer commented on xarray/core/options.py:
```diff
@@ -32,6 +32,7 @@
         "use_numbagg",
         "use_opt_einsum",
         "use_flox",
+        "time_resolution",
```
Agreed, as this switch/parameter will not be used very often in the code, we can come up with a more speaking name. I'd go with your suggestion, if we do not come up with anything better.

kmuehlbauer replied:
Replaced the global option via `time_unit` kwarg in `decode_cf` and descending functions..

kmuehlbauer commented on xarray/core/variable.py:
```diff
@@ -298,11 +246,12 @@ def convert_non_numpy_type(data):
         data = utils.to_0d_object_array(data)
 
     if isinstance(data, pd.Timestamp):
```
That would be great if this could all flow through `_possibly_convert_objects`.

kmuehlbauer replied:
I've moved it accordingly.

kmuehlbauer commented on xarray/core/variable.py:
```diff
@@ -298,11 +246,12 @@ def convert_non_numpy_type(data):
+        data = data.to_numpy()
 
+    if isinstance(data, datetime):
+        data = np.datetime64(data)
     if isinstance(data, timedelta):
-        data = np.timedelta64(getattr(data, "value", data), "ns")
+        data = np.timedelta64(data)
```
I've tried to drop pd.Series here, but failed. So I'm open to any solution.

shoyer replied:
> I think pandas still converts scalar datetime.datetime and datetime.timedelta objects to nanosecond precision values instead of microsecond precision values

This behavior seems quite bad from pandas. Hopefully there's willingness to fix it upstream?

spencerkclark replied:
Yeah, I agree, we should inquire about it if there is not an open issue already.

kmuehlbauer commented on xarray/tests/test_coding_times.py:
```diff
@@ -628,13 +663,14 @@ def test_cf_timedelta_2d() -> None:
     ],
 )
 def test_infer_timedelta_units(deltas, expected) -> None:
+    # todo: why testing, the same thing four times?
```
Ah, copy&pasta error, I see. Yes, should be fixed in main, separately.

shoyer commented on doc/user-guide/time-series.rst:
```diff
@@ -21,20 +21,32 @@ core functionality.
-with numpy and smooth integration with pandas.
+Xarray uses the numpy dtypes ``datetime64[unit]`` and ``timedelta64[unit]``
+(where unit is anything of "s", "ms", "us" and "ns") to represent datetime
+data, which offer vectorized (if sometimes buggy) operations with numpy and
```
Maybe delete `(if sometimes buggy)` now?

The only real bug that I'm familiar with is related to what NumPy considers as the built-in Python scalar corresponding to `ns` precision, which is some form of `int` instead of `datetime.datetime`. This is mitigated significantly by allowing for other datetime64 units.

shoyer commented on doc/user-guide/time-series.rst:
```diff
@@ -68,15 +80,9 @@ You can manual decode arrays in this form by passing a dataset to
-packaged with earlier versions ``netCDF4``).  See :ref:`CFTimeIndex` for more
-information.
+From xarray 2024.11 the resolution of the dates can be tuned between "s", "ms", "us" and "ns". One limitation of using ``datetime64[ns]`` is that it limits the native representation of dates to those that fall between the years 1678 and 2262, which gets increased significantly with lower resolutions. When a netCDF file contains dates outside of these bounds (or dates < 1582-10-15), dates will be  ...
+:py:class:`~xarray.CFTimeIndex` enables a subset of the indexing functionality of a :py:class:`pandas.DatetimeIndex` and is only fully compatible with the standalone version of ``cftime`` (not the version packaged with earlier versions ``netCDF4``).
```
netcdf4-python has not included cftime for at least five year at this point
```suggestion
:py:class:`~xarray.CFTimeIndex` enables a subset of the indexing functionality of a :py:class:`pandas.DatetimeIndex`.
```

shoyer commented on xarray/coding/times.py:
```diff
@@ -363,24 +437,36 @@ def decode_cf_datetime(
-    assert result.dtype == "timedelta64[ns]"
+    unique_timedeltas = np.unique(result[pd.notnull(result)])
+    unit = _netcdf_to_numpy_timeunit(_infer_time_units_from_diff(unique_timedeltas))
+    if unit not in ["s", "ms", "us", "ns"]:
+        unit = "s"
```
Should this raise an error instead?

kmuehlbauer replied:
I think no. `pd.to_timedelta().to_numpy()` returns `timedelta64[ns]`. The following check and astype just make sure, that the result is in the lowest possible resolution but within the range of `TimedeltaIndex` (["s", "ms", "us", "ns"]).

shoyer commented on xarray/coding/times.py:
```diff
@@ -363,24 +437,36 @@ def decode_cf_datetime(
-    result = to_timedelta_unboxed(ravel(num_timedeltas), unit=units)
+    unit = _netcdf_to_numpy_timeunit(units)
+    as_unit = unit
+    if unit not in ["s", "ms", "us", "ns"]:
+        as_unit = "s"
```
same concern as above -- should this raise an error instead?

kmuehlbauer replied:
Similar to above, but the other way round (decoding). `pd_to_timedelta` returns `ns` resolution. This makes sure the data is returned in the lowest possible resolution the intermediate `TimedeltaIndex` can handle. So for decoding "days" we finally get `s` resolution, which seems more appropriate than `ns`.

shoyer commented on xarray/core/options.py:
```diff
@@ -258,6 +267,8 @@ class set_options:
     warn_for_unclosed_files : bool, default: False
         Whether or not to issue a warning when unclosed files are
         deallocated. This is mostly useful for debugging.
+    time_resolution : {"s", "ms", "us", "ns"}, default: "s"
+        Time resolution used for CF encoding/decoding.
```
If possible, we should prefer local control for setting a default time precision.  This would logical to do with encoding, e.g., `encoding['dtype'] = "datetime64[s]"`, or perhaps via keyword argument added to `xarray.decode_cf()`.

Once we do this, do we need this global option?

kmuehlbauer replied:
In the first place I thought the global option would be  a neat solution. But it has also it's drawbacks (like with multiprocessing). I'm working towards your suggestion.

kmuehlbauer replied:
I've removed the global option completely. Currently this decodes to the lowest possible resolution. I'm now checking if this can be applied via keyword argument (most visible approach I think).

kmuehlbauer replied:
@shoyer I've added a `time_unit` kwarg (any suggestion for better naming appreciated) for `decode_cf` and descending functions.

kmuehlbauer replied:
ToDO: Note to self: Is time_resolution still needed at this location?

kmuehlbauer replied:
Removed docstring

shoyer commented on xarray/core/pdcompat.py:
```diff
@@ -73,13 +75,29 @@ def __repr__(self) -> str:
+    dt = pd.Timestamp(*args, **kwargs)
+    units = ["s", "ms", "us", "ns"]
+    default = _get_datetime_resolution()
+    if units.index(default) > units.index(dt.unit):
+        dt = _timestamp_as_unit(dt, default)
```
This is a little surprising. From the docstring of this function, I would expect it to always convert the units -- or perhaps raise an error is that isn't possible.

kmuehlbauer replied:
OK, maybe the naming of the function/docstring is not really appropriate. This should only convert to higher resolutions, if needed. But not the other way round, as this might mean loss of resolution. I'll look into this, thanks!

shoyer commented on xarray/core/variable.py:
```diff
@@ -351,10 +300,13 @@ def _as_array_or_item(data):
-            data = np.timedelta64(data, "ns")
+        kind = data.dtype.kind
+        if kind in "mM":
+            unit = np.datetime_data(data.dtype)[0]
```
For readability, prefer using tuple unpacking:
```suggestion
            unit, _ = np.datetime_data(data.dtype)
```

shoyer reviewed (commented):
Generally, this looks like a big step in the right direction. Thanks for tackling this, Kai!

ChrisBarker-NOAA commented:
Not to throw too much of a wrench in the works here -- so feel free to disregard, but there's an issue I've faced with (single precision) float time encoding: 

Folks (carelessly :-( ) sometimes encode times as "days since ..." using a single precision float. The problem here is not unnecessary precision, as you get with double, but too little -- if you go more than a few years out , you lose seconds precision. (the key problem is that float time -- its precision is a function of the magnitude -- not good for this use case)

The end result is that I get things like model timesteps that are supposed to be hourly, reporting as, e.g. 12:00:18, rather than 12:00:00

One way I've dealt with this is rounding to the minute, or even to hours (if I know the output is hourly), or perhaps  to 10 minutes.

Could / should xarray provide a facility for doing this? maybe?

I guess what I'm proposing is that there be some way to tell xarray to store / save a time variable with e.g. second precision, but to round it to something more coarse when decoding.

*maybe* this could even be automatic / inferred:

if a time is in float days since -- it almost certainly is NOT millisecond precision, or even second precision -- and you could even look at the values (the first one?) and see what the minimum precision is for that timespan.

If I've done my math right, a float can only store second precision for a little over three years. So if the values are greater than three years, you don't have second precision.

Anyway, maybe way too much magic, but it would be nice for my use cases :-)

Example:
```
# 15 min timestep
In [57]: dates
Out[57]: 
[datetime.datetime(2024, 1, 1, 0, 0),
 datetime.datetime(2024, 1, 1, 0, 15),
 datetime.datetime(2024, 1, 1, 0, 30),
 datetime.datetime(2024, 1, 1, 0, 45)]

# common choice of units (though a bad one :-( )
In [58]: units
Out[58]: 'days since 1970-01-01T00:00:00'

# convert to numbers, used float64 by default
In [59]: nums_double = nc4.date2num(dates, units)

# truncate to float32
In [60]: nums_float = nums_double.astype(np.float32)

# convert back to datetimes:
In [61]: dates_float = nc4.num2date(nums_float, units)

In [62]: dates_float
Out[62]: 
array([cftime.DatetimeGregorian(2024, 1, 1, 0, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(2024, 1, 1, 0, 14, 3, 750000, has_year_zero=False),
       cftime.DatetimeGregorian(2024, 1, 1, 0, 30, 56, 250000, has_year_zero=False),
       cftime.DatetimeGregorian(2024, 1, 1, 0, 45, 0, 0, has_year_zero=False)],
      dtype=object)
In [67]: [str(dt) for dt in dates_float]
Out[67]: 
['2024-01-01 00:00:00',
 '2024-01-01 00:14:03.750000',
 '2024-01-01 00:30:56.250000',
 '2024-01-01 00:45:00']
```
Ouch! so what were 15 minute timesteps is now off by about one minute -- and what's too bad is that rounding to the minute wouldn't be right either -- you'd need to round to maybe 5 minutes?

Anyway, maybe this simply isn't xarray's problem to solve -- data providers shouldn't make such mistakes :-(

spencerkclark commented:
@ChrisBarker-NOAA yeah, I agree this kind of situation is annoying, but my feeling is that trying to fix this automatically would be too much magic.  Xarray has convenient functionality for rounding times, which can be used to correct this explicitly—that would be my preference.  E.g. for your example it would look like:
```
>>> decoded
<xarray.DataArray 'time' (time: 4)> Size: 32B
array([cftime.DatetimeGregorian(2024, 1, 1, 0, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(2024, 1, 1, 0, 14, 3, 750000, has_year_zero=False),
       cftime.DatetimeGregorian(2024, 1, 1, 0, 30, 56, 250000, has_year_zero=False),
       cftime.DatetimeGregorian(2024, 1, 1, 0, 45, 0, 0, has_year_zero=False)],
      dtype=object)
Dimensions without coordinates: time
>>> decoded.dt.round("5min")
<xarray.DataArray 'round' (time: 4)> Size: 32B
array([cftime.DatetimeGregorian(2024, 1, 1, 0, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(2024, 1, 1, 0, 15, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(2024, 1, 1, 0, 30, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(2024, 1, 1, 0, 45, 0, 0, has_year_zero=False)],
      dtype=object)
Dimensions without coordinates: time
```

ChrisBarker-NOAA commented:
> Xarray has convenient functionality for rounding times

Oh, nice: I had missed that!  -- you're probably right, too much magic to do for people.

kmuehlbauer commented:
@spencerkclark @ChrisBarker-NOAA I've implemented automated decoding of floating point data to the needed resolution, even when the wanted resolution does not apply. 

Unfortunately the above outlined behaviour is too much involved to be put into the decoder. Nevertheless maybe we can distill some best practices from your vast experience with data @ChrisBarker-NOAA and create a nice example how to handle these difficulties?

ChrisBarker-NOAA commented:
> create a nice example how to handle these difficulties?

Sure -- where would be a good home for that?

kmuehlbauer commented:
Not sure, but https://docs.xarray.dev/en/stable/user-guide/time-series.html could have a dedicated floating point date section.

kmuehlbauer commented:
I've added a kwarg `time_unit` into the `decode_cf` and subsequent functionality.

But instead of adding that kwarg we could slightly overload the `decode_times` to take one of `"s"`, `"ms"`, `"us"`, "ns" with `"ns"` as default.

This would have the positive effect, that we wouldn't need the additional kwarg and have to distribute it through the backends.

- `decode_times=None` - directs to `decode_times=True`
- `decode_times=False` - no decoding 
- `decode_times=True` - decode times with default value ("ns")
- `decode_times="s"` - decode times to at least "s"
- `decode_times="ms"` - decode times to at least "ms"
- `decode_times="us"` - decode times to at least "us"
- `decode_times="ns"` - decode times to "ns"

We could guard `decode_times=None` and `decode_times=True` with a `DeprecationWarning` and add our new defaults in the WarningMessage (eg. "us"). 

This methodology would be fully backwards compatible. It advertises the change via DeprecationWarning in normal operation and also if issues appear in the decoding steps.

If this is something which makes sense @shoyer, @dcherian, @spencerkclark, I'd add the needed changes to this PR.

dcherian commented:
Alternatively, we could make small progress on https://github.com/pydata/xarray/issues/4490 and have 

```python
from xarray.coding import DatetimeCoder

ds = xr.open_mfdataset(..., decode_times=DatetimeCoder(units="ms"))
```

In the long term, it seems nice to have the default use the "natural" units i.e. `"h"` for `units="hours since ..."` and apparently even `"M"` for `units=months since ...` (!!)

> https://numpy.org/doc/stable/reference/arrays.datetime.html#basic-datetimes
> The date units are years (‘Y’), months (‘M’), weeks (‘W’), and days (‘D’), while the time units are hours (‘h’), minutes (‘m’), seconds (‘s’), milliseconds (‘ms’),

kmuehlbauer commented:
> Alternatively, we could make small progress on #4490 and have
> 
> ```python
> from xarray.coding import DatetimeCoder
> 
> ds = xr.open_mfdataset(..., decode_times=DatetimeCoder(units="ms"))
> ```

This took a while to sink in :wink: Yes, that's a neat move. I'll incorporate this suggestion.

> In the long term, it seems nice to have the default use the "natural" units i.e. `"h"` for `units="hours since ..."` and apparently even `"M"` for `units=months since ...` (!!)

As long as we use `pd.Timestamp` for parsing the time unit specification (eg. `seconds since 1992-10-8 15:15:42.5 -6:00`) we can only do this for those units `pd.Timestamp` supports (‘s’, ‘ms’, ‘us’, and ‘ns’). We could add some code that checks if the data can be represented in "days" or "hours" (as given in the time unit specification) and convert after the parsing. Not sure how much there is involved. And this won't work for indexes, as those are restricted to (‘s’, ‘ms’, ‘us’, and ‘ns’).

spencerkclark commented:
+1 for fewer arguments to `open_dataset`, however that is achieved!

Indeed for those practical reasons I do not think it is worth trying to match the on-disk units of integer data any more closely.    Second precision already allows for a time span of roughly +/- 290 billion years (many times older than the Earth), which I think is plenty for most applications :).  

Monthly or yearly units are also somewhat awkward to deal with due to their [different (albeit often violated) definition in the CF conventions](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.11/cf-conventions.html#time-coordinate).

qq492947833 commented:
Same question，the CMIP6 data have 2300 dataset，but now xarray can just read data before 2262.So if you can fix this problem,it will be helpful to us.Thanks a lot!

ChrisBarker-NOAA commented:
> In the long term, it seems nice to have the default use the "natural" units i.e. "h" for units="hours since ..." and apparently even "M" for units=months since ... (!!)

While appealing, I think this is not a good idea. A couple points:

NOTE: I got a bit lost in the discussion, yoy all may have already come to these same conclusions, but I thought I 'd capture it here in one big post ;-)

"months" and "years" are NOT recommended by CF, as they are not clearly defined timespans. (thought UDUNITS does have a definition for them, which is the average -- e.g. 365.25 days to the year (or thereabouts...)

As for days (or even hours):

1) most of the time, these are used with a floating point type anyway.

2) if an integer type -- that does not mean that the unit is not necessarily the required / desired precision.
 - One of the limitations  with the CF encoding of time is that it it's inherently a continuum (to the precision used), and that doesn't change with the units.
 - For instance, a user might *want* daily data (maybe a daily average temperature) or monthly, or ... -- but there is no way to actually express that directly with a CF time [*]. That is:
 `unit: "days since 01-01-2024"`
 `values: [0, 1, 2, 3, 4]`

looks like it's it's expressing Jan 1, 2, 3, 4, ....

But what that actually means is:
```
01-01-2024T00:00:00
01-02-2024T00:00:00
01-03-2024T00:00:00
01-04-2024T00:00:00
...
```
That is, the zeroth time of each day -- i.e. a specific point in the time continuum. 

And this maps to what all (that I know of) the datetime objects do too (python `datetime.date` notwithstanding -- and I don't think it does months)

if a user does want to have a way to express the "day", they might do:

 `unit: "days since 01-01-2024T12:00:00"`
 `values: [0, 1, 2, 3, 4]`

That is, noon of each day.

But then we can't use days as the unit with the fixed epoch of numpy datetime64.

Anyway, all this to say -- I don't think that there is ever a use case for using numpy datetime units longer than a second, certainly not by assuming something from the units of the time.

Using seconds as a default for any encoding of seconds or longer seems reasonable to me, though. But is there any real loss to using milliseconds?



[*] The way to express, e.g. a daily average, is to use "cell bounds", specifically defining the bounds of the average.

kmuehlbauer commented:
Thanks @ChrisBarker-NOAA, I think we should move all these valuable comments in this PR into the docs somehow. I can take a look when this one is finalized.

kmuehlbauer commented:
@dcherian @shoyer @spencerkclark This is ready for another round of review. I'm having a hard time wrt typing, so I'd appreciate any guidance there.

> Alternatively, we could make small progress on #4490 and have
> 
> ```python
> from xarray.coding import DatetimeCoder
> 
> ds = xr.open_mfdataset(..., decode_times=DatetimeCoder(units="ms"))
> ```

I've now implemented that as:

```python
from xarray import CFDatetimeCoder
ds = xr.open_mfdataset(..., decode_times=CFDatetimeCoder(time_unit="ms"))
```
I'm not sure, maybe we should not add `CFDatetimeCoder` to the normal API, but to the advanced API? I could change `time_unit` to `unit`, `units` won't work because that is used in underlaying functions as the `time unit` specification (eg. `seconds since 1992-10-8 15:15:42.5 -6:00`). What a mess :grimacing:.

kmuehlbauer commented:
So the changes added by @ilan-gold over in #9671 help to move things forward in https://github.com/xarray-contrib/xvec/issues/87. It would be great to get another review here and some directions on how to move forward with this. This has come quite far now and there might only be slight changes in naming and API (see my comment https://github.com/pydata/xarray/pull/9618#issuecomment-2503361761). Thanks!

dcherian commented on xarray/conventions.py:
```diff
@@ -191,7 +193,28 @@ def decode_cf_variable(
+
+                warn(
+                    "Usage of 'use_cftime' as kwarg is deprecated. "
+                    "Please initialize it with xarray.CFDatetimeCoder and "
```
reminder to change to `xr.coders.CFDatetimeCoder` or whatever we decide to do.

dcherian reviewed (commented):
We need to have a big note in `whats-new.rst` about behaviour changes.
1. The default for `pd.date_range` seems to have changed, and we now preserve the unit instead of casting. This will impact downstream code.
2. Has the default unit for `xr.date_range` changed? We'll need to add the `unit` kwarg with default `ns` and start warning about switching to `us` by default.
3. What is the impact on `polyfit` and `differentiate`? These are places where I (and many others) have manually rescaled from `ns` to other units dividing by `1e9`. We don't want to break this silently.

kmuehlbauer commented on doc/user-guide/time-series.rst:
```diff
@@ -51,7 +62,7 @@ attribute like ``'days since 2000-01-01'``).
 
    When decoding/encoding datetimes for non-standard calendars or for dates
-   before year 1678 or after year 2262, xarray uses the `cftime`_ library.
+   before 1582-10-15, xarray uses the `cftime`_ library.
```
I'm adding this suggestion, but we might want to note that this holds true only for non-nanosecond data.

kmuehlbauer commented on doc/user-guide/weather-climate.rst:
```diff
@@ -75,18 +74,14 @@ using a standard calendar, but outside the `nanosecond-precision range`_
    - The dates are from a non-standard calendar
-   - Any dates are outside the nanosecond-precision range.
+   - Any dates are outside the nanosecond-precision range (prior xarray version 2024.11)
+   - Any dates are outside the time span limited by the resolution (from xarray version v2024.11)
```
TODO: Fix version

kmuehlbauer replied:
```suggestion
   - Any dates are outside the time span limited by the resolution (from xarray version v2025.01.1)
```

kmuehlbauer reviewed (commented):
Thank you for the review Deepak, :heart: 

I'll try to get along over the weekend.

kmuehlbauer commented on doc/user-guide/weather-climate.rst:
```diff
@@ -75,18 +74,14 @@ using a standard calendar, but outside the `nanosecond-precision range`_
-   represented with the ``np.datetime64[ns]`` data type, enabling the use of a
-   :py:class:`pandas.DatetimeIndex` or arrays with dtype ``np.datetime64[ns]``
-   and their full set of associated features.
+   represented with the ``np.datetime64[unit]`` data type (where unit can be any of ["s", "ms", "us", "ns"], enabling the use of a :py:class:`pandas.DatetimeIndex` or arrays with dtype ``np.datetime64[unit]`` and their full set of associated features.
```
```suggestion
   represented with the ``np.datetime64[unit]`` data type (where unit can be one of ["s", "ms", "us", "ns"], enabling the use of a :py:class:`pandas.DatetimeIndex` or arrays with dtype ``np.datetime64[unit]`` and their full set of associated features.
```

kmuehlbauer commented:
> I see many notes about the units s-us. Will this work with minutes or days too? Eg date_range and other sources

@alippai Can you give an example what you have in mind? Currently xarray holds datetimes as `np.datetime64[ns]` under the hood. This PR relaxes this restriction so datetimes can be represented as `np.datetime64[s]`, `np.datetime64[ms]` and `np.datetime64[us]`, too.

kmuehlbauer commented:
> We need to have a big note in `whats-new.rst` about behaviour changes.
> 
>     1. The default for `pd.date_range` seems to have changed, and we now preserve the unit instead of casting. This will impact downstream code.
> 
>     2. Has the default unit for `xr.date_range` changed? We'll need to add the `unit` kwarg with default `ns` and start warning about switching to `us` by default.
> 
>     3. What is the impact on `polyfit` and `differentiate`? These are places where I (and many others) have manually rescaled from `ns` to other units dividing by `1e9`. We don't want to break this silently.

Thanks @dcherian for bringing this to attention. I think we can circumvent these behaviour changes. Let me try first.

alippai commented:
> > I see many notes about the units s-us. Will this work with minutes or days too? Eg date_range and other sources
> 
> @alippai Can you give an example what you have in mind? Currently xarray holds datetimes as `np.datetime64[ns]` under the hood. This PR relaxes this restriction so datetimes can be represented as `np.datetime64[s]`, `np.datetime64[ms]` and `np.datetime64[us]`, too.

Larger units like np.datetime64[D] and m types

kmuehlbauer commented on doc/internals/time-coding.rst:
```diff
@@ -0,0 +1,444 @@
+    print(values, units)
+
+
+Default Time Unit
```
Resolved

kmuehlbauer commented:
@dcherian @spencerkclark I'm now trying to split out everything which helps to get the footprint lower here:

- #9882 
- #9885 
- #9899 
- use iso-parser when reference time out-of-bounds 
- move scalar handling logic into `_possibly_convert_objects`
- split out CFDatetimeCoder/ deprecate use_cftime kwarg
- tbc

kmuehlbauer commented on doc/user-guide/time-series.rst:
```diff
@@ -66,17 +77,15 @@ You can manual decode arrays in this form by passing a dataset to
-fully compatible with the standalone version of ``cftime`` (not the version
-packaged with earlier versions ``netCDF4``).  See :ref:`CFTimeIndex` for more
-information.
+From xarray TODO: version the resolution of the dates can be tuned between "s", "ms", "us" and "ns". One limitation of using ``datetime64[ns]`` is that it limits the native representation of dates to those that fall between the years 1678 and 2262, which gets increased significantly with lower resolutions. When a netCDF file contains dates outside of these bounds (or dates < 1582-10-15), dates wi ...
```
```suggestion
From xarray TODO: version the resolution of the dates can be tuned between "s", "ms", "us" and "ns". One limitation of using ``datetime64[ns]`` is that it limits the native representation of dates to those that fall between the years 1678 and 2262, which gets increased significantly with lower resolutions. When a netCDF file contains dates outside of these bounds (or dates < 1582-10-15) and no 'proleptic_gregorian' calendar is used, dates will be returned as arrays of :py:class:`cftime.datetime` objects and a :py:class:`~xarray.CFTimeIndex` will be used for indexing.
```

kmuehlbauer commented on doc/user-guide/weather-climate.rst:
```diff
@@ -75,18 +74,14 @@ using a standard calendar, but outside the `nanosecond-precision range`_
 
    - The dates are from a non-standard calendar
-   - Any dates are outside the nanosecond-precision range.
+   - Any dates are outside the nanosecond-precision range (prior xarray version 2024.11)
```
```suggestion
   - Any dates are outside the nanosecond-precision range (prior xarray version 2025.01.1)
```

kmuehlbauer commented on doc/user-guide/weather-climate.rst:
```diff
@@ -75,18 +74,14 @@ using a standard calendar, but outside the `nanosecond-precision range`_
-   represented with the ``np.datetime64[ns]`` data type, enabling the use of a
-   :py:class:`pandas.DatetimeIndex` or arrays with dtype ``np.datetime64[ns]``
-   and their full set of associated features.
+   represented with the ``np.datetime64[unit]`` data type (where unit can be one of ["s", "ms", "us", "ns"], enabling the use of a :py:class:`pandas.DatetimeIndex` or arrays with dtype ``np.datetime64[unit]`` and their full set of associated features.
```
```suggestion
   represented with the ``np.datetime64[unit]`` data type (where unit can be one of ["s", "ms", "us", "ns"]), enabling the use of a :py:class:`pandas.DatetimeIndex` or arrays with dtype ``np.datetime64[unit]`` and their full set of associated features.
```

kmuehlbauer commented on doc/user-guide/weather-climate.rst:
```diff
@@ -75,18 +74,14 @@ using a standard calendar, but outside the `nanosecond-precision range`_
-   to nanosecond-precision for backwards compatibility with older pandas
-   versions; however, this is something we would like to relax going forward.
-   See :issue:`7493` for more discussion.
+   values. From xarray version 2024.11 the relaxed non-nanosecond precision datetime values will be used.
```
```suggestion
   values. From xarray version 2025.01.1 relaxed non-nanosecond precision datetime values can be parameterized via :py:class:`coders.CFDatetimeCoder` and ``decode_times` kwarg.
```

kmuehlbauer commented on doc/whats-new.rst:
```diff
@@ -86,6 +88,10 @@ Breaking changes
 
 Deprecations
 ~~~~~~~~~~~~
+- Time decoding related kwarg ``use_cftime`` is deprecated. Use keyword argument
+  ``decode_times=CFDatetimeCoder(use_cftime=True)`` in the respective functions
+  instead (:pull:`9618`).
+  By `Kai Mühlbauer <https://github.com/kmuehlbauer>`_.
```
```suggestion
```

kmuehlbauer commented on doc/whats-new.rst:
```diff
@@ -72,7 +74,7 @@ New Features
 - Add ``unit`` - keyword argument to :py:func:`date_range` and ``microsecond`` parsing to
   iso8601-parser (:pull:`9885`).
-  By `Kai Mühlbauer <https://github.com/kmuehlbauer>`_.
+  By `Kai Mühlbauer <https://github.com/kmuehlbauer>`_
```
```suggestion
  By `Kai Mühlbauer <https://github.com/kmuehlbauer>`_.
```

kmuehlbauer commented on xarray/conventions.py:
```diff
@@ -155,6 +155,11 @@ def decode_cf_variable(
         represented using ``np.datetime64[ns]`` objects.  If False, always
         decode times to ``np.datetime64[ns]`` objects; if this is not possible
         raise an error.
+        Usage of use_cftime as kwarg is deprecated, please initialize it with
+        CFDatetimeCoder and ``decode_times``.
+
+        .. deprecated:: 2025.01.0
+           Please pass a :py:class:`coders.CFDatetimeCoder` instance initialized with ``use_cftime`` to the ``decode_times`` kwarg instead.
```
```suggestion
```

kmuehlbauer commented:
I think I've fixed up all issues due to manually merging. Hopefully this is now in a consistent state. I'm not sure if this is good enough, or if I should just rebase/squash this on latest master? Any thoughts appreciated.

shoyer commented on xarray/core/variable.py:
```diff
@@ -205,51 +204,28 @@ def _maybe_wrap_data(data):
     as_series = pd.Series(values.ravel(), copy=False)
-    if as_series.dtype.kind in "mM":
-        as_series = _as_nanosecond_precision(as_series)
+    # When receiving objects which pd.Series can't resolve by its own
+    # we try astype-conversion to "ns"-resolution for datetimes and pd.Timestamp.
+    if (
+        values.dtype.kind == "O"
+        and as_series.dtype.kind == "O"
+        and as_series.size > 0
+        and (
+            isinstance(as_series[0], datetime | pd.Timestamp)
+            or pd.api.types.is_datetime64_dtype(as_series[0])
+        )
+    ):
+        as_series = as_series.astype("=M8[ns]")
```
This part feels a little strange to me and possibly unnecessary. When did this come up?

kmuehlbauer replied:
I'm not sure, when this came in. I'll remove it for now, as it does not seem to have any effect atm. If the issue resurfaces which should be resolved by this construct we can reiterate. Thanks for spotting this.

shoyer commented:
> I think I've fixed up all issues due to manually merging. Hopefully this is now in a consistent state. I'm not sure if this is good enough, or if I should just rebase/squash this on latest master? Any thoughts appreciated.

No need to rebase/squash manually! As long as this PR can be merged in GitHub, it will be squashed automatically

kmuehlbauer commented on xarray/coding/times.py:
```diff
@@ -356,8 +365,49 @@ def _check_date_for_units_since_refdate(
+    # proleptic_gregorian and standard/gregorian are only equivalent
+    # if reference date and date range is >= 1582-10-15
+    if calendar != "proleptic_gregorian":
+        if date < type(date)(1582, 10, 15):
```
I see, yes, the function also consumes `cftime.datetime` (eg. `cftime.DatetimeNoLeap` via `_is_numpy_compatible_time_range`:

https://github.com/kmuehlbauer/xarray/blob/9bd292a0d77c54febd31819cff54e3d0a8166391/xarray/coding/times.py#L98-L115

I'll have to update the typing in the function arguments.

kmuehlbauer replied:
If someone of the typing people could help out here? I've no idea how to do this properly.

Illviljan replied:
Try out using:
https://github.com/pydata/xarray/blob/5279bd15708e21ea582e56e4f27fc3e9663a8509/xarray/core/types.py#L113

kmuehlbauer replied:
Thanks @Illviljan for the hint. I can't use `DatetimeLike` because `np.datetime64` can't be initialized with (year, month, day). But I've just used the definition without `np.datetime64`. This somehow makes the whole thing work.

kmuehlbauer commented on xarray/coding/times.py:
```diff
@@ -356,8 +365,49 @@ def _check_date_for_units_since_refdate(
+        flat_num_dates, new_time_unit = _check_higher_resolution(
+            flat_num_dates * 1000,
+            iter_unit=iter_unit,
+        )
```
Yes, keep it simple. Done.

spencerkclark reviewed (commented):
Thanks @kmuehlbauer for updating this—I tried to give it a first thorough look through.  Some small suggestions, a few possible minor remaining issues, but I think the big picture is looking good!

kmuehlbauer commented on doc/user-guide/weather-climate.rst:
```diff
@@ -75,18 +74,14 @@ using a standard calendar, but outside the `nanosecond-precision range`_
-   to nanosecond-precision for backwards compatibility with older pandas
-   versions; however, this is something we would like to relax going forward.
-   See :issue:`7493` for more discussion.
+   values. From xarray version 2024.11 the relaxed non-nanosecond precision datetime values will be used.
```
Done.

kmuehlbauer commented on doc/user-guide/weather-climate.rst:
```diff
@@ -137,7 +132,7 @@ Conversion between non-standard calendar and to/from pandas DatetimeIndexes is
 :py:meth:`xarray.DataArray.convert_calendar`). Here, like elsewhere in xarray, the ``use_cftime``
 argument controls which datetime backend is used in the output. The default (``None``) is to
-use ``pandas`` when possible, i.e. when the calendar is standard and dates are within 1678 and 2262.
+use ``pandas`` when possible, i.e. when the calendar is ``standard``/``gregorian`` and dates [starting with 1582-10-15]((https://en.wikipedia.org/wiki/Gregorian_calendar)). There is no such restriction when converting to ``proleptic_gregorian`` calendar.
```
Done.

kmuehlbauer commented on doc/internals/time-coding.rst:
```diff
@@ -0,0 +1,442 @@
+
+The function :py:func:`pandas.to_timedelta` is used within xarray for inferring units and for testing purposes.
+
+In normal operation :py:func:`pandas.to_timedelta` returns :py:class:`pandas.Timedelta` (scalar input) or :py:class:`pandas.TimedeltaIndex` (array-like input) which are timedelta64 with ``ns`` resolution internally. That has the implication, that the usable timedelta covers only roughly 585 years. To accommodate for that, we are working around that limitation in the encoding and decoding step.
```
I think I was referring to:


```python
pd.Timedelta(np.iinfo("int64").max, "us")
# or
pd.to_timedelta(np.iinfo("int64").max, "us")
```
It seems pandas can handle the numpy.timedelta64 with their given resolution.

kmuehlbauer commented on xarray/tests/test_concat.py:
```diff
@@ -317,7 +317,7 @@ def test_concat_multiple_datasets_with_multiple_missing_variables() -> None:
 
 
-@pytest.mark.filterwarnings("ignore:Converting non-nanosecond")
+@pytest.mark.filterwarnings("ignore:Converting non-default")
```
Let's check...

kmuehlbauer commented on xarray/core/variable.py:
```diff
@@ -78,13 +78,11 @@
     "np.timedelta64 values to the DataArray or Variable constructor; it can be "
-    "silenced by converting the values to nanosecond precision ahead of time."
+    "silenced by converting the values to default precision {res!r} ahead of time."
 )
```
Removed.

kmuehlbauer commented on xarray/tests/test_variable.py:
```diff
@@ -201,24 +200,25 @@ def test_index_0d_string(self):
+        # todo: discussion needed
+        td64 = np.timedelta64(td, "ns")
+        x = self.cls(["x"], [td64])
+        self._assertIndexedLikeNDArray(x, td64, np.dtype("timedelta64[ns]"))
```
Had to fix scalar index handling to make this work.

kmuehlbauer commented on xarray/tests/test_variable.py:
```diff
@@ -275,52 +275,60 @@ def test_0d_time_data(self):
-    @pytest.mark.filterwarnings("ignore:Converting non-nanosecond")
+    @pytest.mark.filterwarnings("ignore:Converting non-default")
     def test_datetime64_valid_range(self):
+        # todo: test still needed?
```
Done.

kmuehlbauer commented on xarray/tests/test_variable.py:
```diff
@@ -275,52 +275,60 @@ def test_0d_time_data(self):
-    @pytest.mark.filterwarnings("ignore:Converting non-nanosecond")
+    @pytest.mark.filterwarnings("ignore:Converting non-default")
     def test_timedelta64_valid_range(self):
+        # todo: test still needed?
```
Done.

kmuehlbauer commented on xarray/tests/test_variable.py:
```diff
@@ -275,52 +275,60 @@ def test_0d_time_data(self):
+        ],
+    )
+    def test_timedelta64_conversion(self, values, unit):
+        # todo: check, if this test is OK
```
Done.

kmuehlbauer commented on xarray/tests/test_variable.py:
```diff
@@ -1073,31 +1081,38 @@ def test_numpy_same_methods(self):
+        ],
+    )
+    def test_datetime64_conversion_scalar(self, values, unit):
+        # todo: check, if this test is OK
```
Done.

kmuehlbauer commented on xarray/tests/test_variable.py:
```diff
@@ -1073,31 +1081,38 @@ def test_numpy_same_methods(self):
+        ],
+    )
+    def test_timedelta64_conversion_scalar(self, values, unit):
+        # todo: discussion needed
+        # todo: check, if this test is OK
```
Done.

kmuehlbauer commented on xarray/tests/test_variable.py:
```diff
@@ -275,52 +275,60 @@ def test_0d_time_data(self):
+        ],
+    )
+    def test_datetime64_conversion(self, values, unit):
+        # todo: check, if this test is OK
```
Done.

kmuehlbauer commented on xarray/tests/test_variable.py:
```diff
@@ -2976,37 +2997,32 @@ def test_from_pint_wrapping_dask(self, Var):
 )
-def test_datetime_conversion_warning(values, warns) -> None:
+def test_datetime_conversion_warning(values, unit) -> None:
+    # todo: needs discussion
+    # todo: check, if this test is OK
```
I've rephrased the todo and will grep things later.

kmuehlbauer commented on xarray/tests/test_variable.py:
```diff
@@ -3044,65 +3060,34 @@ def test_datetime_conversion_warning(values, warns) -> None:
 ) -> None:
-    with pytest.warns(UserWarning, match="non-nanosecond precision datetime"):
-        var = Variable(["time"], data.astype(dtype))  # type: ignore[arg-type]
+    # todo: check, if this test is OK
```
I've rephrased the todo and will grep things later.

kmuehlbauer commented on xarray/tests/test_variable.py:
```diff
@@ -3044,65 +3060,34 @@ def test_datetime_conversion_warning(values, warns) -> None:
     ids=lambda x: f"{x}",
 )
-def test_timedelta_conversion_warning(values, warns) -> None:
+def test_timedelta_conversion_warning(values, unit) -> None:
```
Added todo for now.

kmuehlbauer commented on xarray/tests/test_coding_times.py:
```diff
@@ -266,30 +272,37 @@ def test_decode_dates_outside_timestamp_range(calendar) -> None:
     assert all(isinstance(value, expected_date_type) for value in actual)
     abs_diff = abs(actual - expected)
     # once we no longer support versions of netCDF4 older than 1.1.5,
     # we could do this check with near microsecond accuracy:
     # https://github.com/Unidata/netcdf4-python/issues/355
-    assert (abs_diff <= np.timedelta64(1, "s")).all()
+    assert (abs_diff <= np.timedelta64(1, "us")).all()
```
Yes, that would be great.

kmuehlbauer commented on xarray/tests/test_coding_times.py:
```diff
@@ -595,8 +624,9 @@ def test_infer_cftime_datetime_units(calendar, date_args, expected) -> None:
     ],
 )
 def test_cf_timedelta(timedeltas, units, numbers) -> None:
+    # todo: check, if this test is OK
```
I've added a fix. Please check if this in line with your comment.

spencerkclark replied:
Almost—I realize my comments about enabling non-nanosecond timedelta decoding, but defaulting to nanosecond resolution decoding could only be satisfied by implementing a `time_unit` option in `encode_cf_timedelta`...I gave it a try in https://github.com/kmuehlbauer/xarray/pull/3.  

Let me know if that looks OK. Your refactoring of the timedelta handling in 03086723dd026970660f8d671575bd54bbef619a really does the heavy lifting 👍.

kmuehlbauer commented on xarray/tests/test_coding_times.py:
```diff
@@ -177,7 +176,7 @@ def test_decode_cf_datetime_overflow() -> None:
             warnings.filterwarnings("ignore", "Unable to decode time axis")
-            result = decode_cf_datetime(day, units)
+            result = decode_cf_datetime(day, units, time_unit=time_unit)
         assert result == expected[i]
```
Done.

kmuehlbauer commented on xarray/coding/times.py:
```diff
@@ -473,36 +554,57 @@ def decode_cf_datetime(
+        as_unit = "s"
+    result = (
+        pd.to_timedelta(ravel(num_timedeltas), unit=unit).as_unit(as_unit).to_numpy()
+    )
```
Yes, this was the reason. I'll re-evaluate.

kmuehlbauer replied:
I've refactored out the timedelta handling and re-use it here.

kmuehlbauer commented on xarray/coding/times.py:
```diff
@@ -881,14 +970,17 @@ def _eagerly_encode_cf_datetime(
 
         time_unit, ref_date = _unpack_time_unit_and_ref_date(units)
+        # calendar equivalence only for days after the reform
+        _check_date_is_after_shift(ref_date, calendar)
```
OK, I think the issue is with having out-of-bounds datetime64 with erroneous standard-calendar. This can't happen when decoding. So we would need to raise in this case.

kmuehlbauer replied:
I'll add the needed construct there.

kmuehlbauer replied:
I've found a slight inconsistency with cftime, which influences a proper solution. See https://github.com/Unidata/cftime/issues/354

kmuehlbauer replied:
This is now fixed with the latest commit. Added a warning for those cases and fallback to cftime encoding.

spencerkclark replied:
Thanks @kmuehlbauer—sorry for sending you down another rabbit hole.  This is a clever solution.

kmuehlbauer commented on doc/internals/time-coding.rst:
```diff
@@ -0,0 +1,442 @@
+
+    time = pd.to_datetime(np.datetime64(0, "D"))
+    print("Datetime:", time, np.asarray([time.to_numpy()]).dtype)
+    print("Datetime as_unit('s'):", time.as_unit("s"))
+    print("Datetime to_numpy():", time.as_unit("s").to_numpy())
```
I've added some context and fixed up the example.

kmuehlbauer commented on doc/internals/time-coding.rst:
```diff
@@ -0,0 +1,442 @@
+    print("TimedeltaIndex to_numpy():", delta.as_unit("s").to_numpy())
+
+.. note::
+    For the functionality in xarray the output resolution is converted from ``'ns'`` to the lowest needed resolution.
```
I've added "when decoding" to the end of that sentence.

kmuehlbauer commented on doc/internals/time-coding.rst:
```diff
@@ -0,0 +1,442 @@
+:py:class:`pandas.DatetimeIndex` will raise :py:class:`pandas.OutOfBoundsDatetime` if the input can't be represented in the given resolution.
+
+.. note::
+    For xarray we assume that all :py:class:`numpy.datetime64` provided to :py:class:`pandas.DatetimeIndex` are up to the specs. This is especially true, when those values have been decoded upfront. If the data is provided by users, they should handle any issues before.
```
Yes, propably not needed. I'll remove it.

dcherian commented on doc/user-guide/time-series.rst:
```diff
@@ -66,17 +77,15 @@ You can manual decode arrays in this form by passing a dataset to
-fully compatible with the standalone version of ``cftime`` (not the version
-packaged with earlier versions ``netCDF4``).  See :ref:`CFTimeIndex` for more
-information.
+From xarray 2025.01.1 the resolution of the dates can be tuned between "s", "ms", "us" and "ns". One limitation of using ``datetime64[ns]`` is that it limits the native representation of dates to those that fall between the years 1678 and 2262, which gets increased significantly with lower resolutions. When a store contains dates outside of these bounds (or dates < 1582-10-15 with a Gregorian, al ...
```
```suggestion
From xarray 2025.01.1 the resolution of the dates can be one of "s", "ms", "us" or "ns". One limitation of using ``datetime64[ns]`` is that it limits the native representation of dates to those that fall between the years 1678 and 2262, which gets increased significantly with lower resolutions. When a store contains dates outside of these bounds (or dates < 1582-10-15 with a Gregorian, also known as standard, calendar), dates will be returned as arrays of :py:class:`cftime.datetime` objects and a :py:class:`~xarray.CFTimeIndex` will be used for indexing.
```

minor wording suggestion

dcherian reviewed (approved):
Amazing stuff, Kai!

kmuehlbauer commented:
@dcherian Thanks for taking another look here. I'm deep in the weeds to iron out the last remaining issues. Hope to fix this up until next week. :crossed_fingers:

kmuehlbauer commented on xarray/core/indexing.py:
```diff
@@ -1754,7 +1754,10 @@ def _convert_scalar(self, item):
-            item = np.timedelta64(getattr(item, "value", item), "ns")
+            # from xarray 2025.01.1 xarray allows non-nanosecond resolution
+            # so we just convert to_numpy if possible
+            if hasattr(item, "to_numpy"):
```
My indexing knowledge really lacks, so any suggestion would help. This construct was just to get the machinery running again.

kmuehlbauer replied:
I've introduced a pd.Timestamp check...

dcherian commented on xarray/coding/cftimeindex.py:
```diff
@@ -644,13 +644,14 @@ def to_datetimeindex(self, unsafe=False):
         CFTimeIndex([2000-01-01 00:00:00, 2000-01-02 00:00:00],
                     dtype='object', length=2, calendar='standard', freq=None)
         >>> times.to_datetimeindex()
```
flagging this too, perhaps a separate PR after this goes in?

kmuehlbauer replied:
Yes, I'm hesitant to add more things to this PR.

kmuehlbauer commented on xarray/tests/test_backends.py:
```diff
@@ -613,6 +613,12 @@ def test_roundtrip_cftime_datetime_data(self) -> None:
+                    # proleptic gregorian will be decoded into numpy datetime64
+                    # fixing to expectations
+                    if actual.t.dtype.kind == "M":
+                        dtype = f"datetime64[{np.datetime_data(actual.t)[0]}]"
```
Let's try... Thanks!

kmuehlbauer commented on xarray/coding/times.py:
```diff
@@ -1165,6 +1266,7 @@ def decode(self, variable: Variable, name: T_Name = None) -> Variable:
 
             units = pop_to(attrs, encoding, "units")
             transform = partial(decode_cf_timedelta, units=units)
+            # todo: check, if we can relax this one here, too
```
Might be better to postpone it to a follow-up PR. It looks like there is more involved here.

kmuehlbauer commented on xarray/coding/times.py:
```diff
@@ -594,36 +691,32 @@ def infer_timedelta_units(deltas) -> str:
-            dt = nanosecond_precision_timestamp(
-                t.year, t.month, t.day, t.hour, t.minute, t.second, t.microsecond
-            )
+            # We expect either "us" resolution or "s" resolution depending on
+            # whether 'microseconds' are defined for the input or not.
+            dt = np.datetime64(t.isoformat()).astype(f"=M8[{time_unit}]")
```
I've rewritten the line like this:

```python
dt = pd.Timestamp(np.datetime64(t.isoformat())).as_unit(time_unit).to_numpy()
```

This should make sure to raise accordingly when overflowing in case of ns resolution.

kmuehlbauer commented on doc/internals/time-coding.rst:
```diff
@@ -0,0 +1,440 @@
+.. ipython:: python
```
I've started with one reference. We can always add further references, when needed.

kmuehlbauer commented:
Slowly approaching the gates...

I _think_ I've addressed all suggestions and review comments. Good chance that I've missed something, though.

One final unresolved issue is with polyfit/integration (cc @dcherian). There seems to be some hardcoding of "ns". Unfortunately this is not my home turf, so maybe someone more experienced should take this on.

The current state of this PR should be backward compatible wrt to time decoding. There are slight changes (and also enhancements/bug fixes). All non-nano enhancements are opt-in (`time_unit`-kwarg to `xr.coders.CFDatetimeCoder`. I've left todo's in the code where to pick up in future PR's. If one of those todo's should be fixed in this PR, please have another review.

I'll need to summarize the changes in the whats-new.rst.

shoyer commented:
> One final unresolved issue is with polyfit/integration (cc @dcherian). There seems to be some hardcoding of "ns". Unfortunately this is not my home turf, so maybe someone more experienced should take this on.

As long as the unit tests are passing, let's open an issue for this and leave it for later. This PR is big enough already. It would be best to continue this work in smaller follow-on PRs.

My cursory examination of `datetime_to_numeric()` (called by `interp` and `integrate` internally) is that it should be totally with different datetime units. In general we only have to worry about extremely unsafe cases where code assumes a datetime array already has `ns` precision without checking.

dcherian commented:
Stephan's right, we force `ns` when converting to numeric for polyfit/polyval so this PR isn't breaking things
https://github.com/pydata/xarray/blob/ceb5364be999567d08ba9be832fa0440e6dd3407/xarray/core/computation.py#L2117

I pushed a couple of extra tests to make sure we don't break this in the future.

kmuehlbauer commented on doc/user-guide/time-series.rst:
```diff
@@ -21,20 +21,31 @@ core functionality.
 
 .. ipython:: python
 
     pd.to_datetime(["2000-01-01", "2000-02-02"])
```
Added xr.date_range example

kmuehlbauer commented:
I went over the docs once again, also added whats-new.rst entry. Some more cosmetic changes. I'd say this is finished. Please let me know, if there is something missing.

spencerkclark reviewed (approved):
Thank you for all of your hard work on this @kmuehlbauer 🙏.

I think we are basically there!  I just want to make sure we are not committing anything that we know is a breaking change or otherwise does work out of the box.  I'm approving, but take a look at https://github.com/kmuehlbauer/xarray/pull/3, which fixes a couple issues related to timedelta decoding / encoding.

kmuehlbauer commented:
Thanks again @spencerkclark! I've merged your https://github.com/kmuehlbauer/xarray/pull/3 as it is essential for this PR to keep consistency and backwards compatibility.

kmuehlbauer commented on doc/whats-new.rst:
```diff
@@ -19,9 +19,39 @@ What's New
 ~~~~~~~~~~~~
-
+- Relax nanosecond datetime restriction in CF time decoding (:issue:`7493`, :pull:`9618`).
+  By `Kai Mühlbauer <https://github.com/kmuehlbauer>`_.
```
I'd really like to see @spencerkclark mentioned here. This would not have been possible with your continuing support, suggestions and additions.

```suggestion
  By `Kai Mühlbauer <https://github.com/kmuehlbauer>`_ and `Spencer Clark <https://github.com/spencerkclark>`_.
```

spencerkclark replied:
I appreciate it Kai—I'm glad I could help here and there!  Again, amazing work.

spencerkclark commented on xarray/tests/test_coding_times.py:
```diff
@@ -615,20 +651,53 @@ def test_cf_timedelta(timedeltas, units, numbers) -> None:
+    assert result.dtype == expected.dtype
+
+
+def test_decode_cf_timedelta_time_unit_out_of_bounds(time_unit):
```
Realized I forgot the return types here and below:
```suggestion
def test_decode_cf_timedelta_time_unit_out_of_bounds(time_unit) -> None:
```

spencerkclark commented on xarray/tests/test_coding_times.py:
```diff
@@ -615,20 +651,53 @@ def test_cf_timedelta(timedeltas, units, numbers) -> None:
+        decode_cf_timedelta(encoded, "days", time_unit)
+
+
+def test_cf_timedelta_roundtrip_large_value(time_unit):
```
```suggestion
def test_cf_timedelta_roundtrip_large_value(time_unit) -> None:
```

kmuehlbauer commented:
Failing CI looks unrelated. Decision in the dev meeting was to merge fast to have a bit of time to test before release later this month.

kmuehlbauer commented:
Thanks All! :rocket:

spencerkclark commented:
Awesome, congrats @kmuehlbauer—thank *you*!

ilan-gold commented:
Yes @kmuehlbauer - monumental effort!!!

dcherian commented:
Amazing! :clap: :clap: :clap:
